cpp/recognition/cograph_recognition/Recognition.cpp

```diff
@@ -0,0 +1,331 @@
+                }
+                auto *z = new CoNode(Type::ZERO_ONE, 0);
+                T->Add(z);
+                y->AddChild(z);
+                z->AddChild(x);
+                z->AddChild(u);
+            } else {
+                auto *z = new CoNode(Type::ZERO_ONE, 1);
+                T->Add(z);
+                u->AddChild(z);
+                z->AddChild(x);
+                z->AddChild(y);
+            }
+        }
+    }
+
+
+    CorneilStewartPerlCographRecognition::State
+    CorneilStewartPerlCographRecognition::Cograph_Recognition(NetworKit::Graph &graph) {
+        auto *R = new CoNode(Type::ZERO_ONE, 1);
+        CoTree Tp(R);
+        T = &Tp;
+        G = graph;
+        std::vector<NetworKit::node> vertex;
+        std::vector<CoNode *> covertex;
+        std::map<NetworKit::node, int> pos;
+        int cnt = 0;
+        for (auto i: G.nodeRange()) {
+            vertex.push_back(i);
+            pos[i] = cnt++;
+            auto *C = new CoNode(Type::VERTEX, int(i));
+            T->Add(C);
+            covertex.push_back(C);
+        }
+        for (auto i: G.nodeRange()) {
+            std::vector<CoNode *> vec;
+            for (auto u: G.neighborRange(i)) {
+                vec.push_back(covertex[pos[u]]);
+            }
+            covertex[pos[i]]->out_edges = vec;
+        }
+
+        if (cnt == 0) {
+            T->Clear();
+            return State::COGRAPH;
+        }
+        if (cnt == 1) {
+            R->AddChild(covertex[0]);
+            T->Clear();
+            return State::COGRAPH;
+        }
+        if (G.hasEdge(vertex[0], vertex[1])) {
+            R->AddChild(covertex[0]);
+            R->AddChild(covertex[1]);
+        } else {
+            auto *N = new CoNode(Type::ZERO_ONE, 0);
+            T->Add(N);
+            R->AddChild(N);
+            N->AddChild(covertex[0]);
+            N->AddChild(covertex[1]);
+        }
+        covertex[0]->in_graph = true;
+        covertex[1]->in_graph = true;
+        CoNode *root;
+
+        for (int i = 2; i < cnt; i++) {
+            root = T->root;
+            ResetAllCoNodes(root);
+            Mark(covertex[i]);
+            if (root->marked ==
+                Marked::MARKED_AND_UNMARKED) {//all nodes of T were marked and unmarked <=> R is marked and unmarked
+                root->AddChild(covertex[i]);
+            } else if (mark_ever_count == 0) {
+                if (root->d == 1) {
+                    (root->head_of_list_of_children)->AddChild(covertex[i]);
+                } else {
+                    auto *R1 = new CoNode(Type::ZERO_ONE, 1);
+                    auto *R2 = new CoNode(Type::ZERO_ONE, 0);
+                    T->Add(R1);
+                    T->Add(R2);
+                    R1->AddChild(R2);
+                    R2->AddChild(root);
+                    R2->AddChild(covertex[i]);
+                    T->root = R1;
+                }
+            } else {
+                auto u = FindLowest();
```

Comment:
You can unpack the result right here: `auto vertex, state = FindLowest();`

Reply:
fixed